python-checks/src/main/java/org/sonar/python/checks/TfFunctionRecursivityCheck.java

```diff
@@ -0,0 +1,114 @@
+import java.util.Set;
+import javax.annotation.Nonnull;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.symbols.Usage;
+import org.sonar.plugins.python.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Decorator;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.FunctionDef;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6908")
+public class TfFunctionRecursivityCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Remove this recursive call.";
+  private static final String SECONDARY_MESSAGE = "Recursive call is here.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FUNCDEF, TfFunctionRecursivityCheck::checkFunctionDef);
+  }
+
+  private static void checkFunctionDef(SubscriptionContext context) {
+    FunctionDef functionDef = (FunctionDef) context.syntaxNode();
+    if (!isTfFunction(functionDef)) {
+      return;
+    }
+    var selfSymbol = TreeUtils.getFunctionSymbolFromDef(functionDef);
+    if (selfSymbol == null) {
+      return;
+    }
+    var collector = new CallCollector(selfSymbol);
+    context.syntaxNode().accept(collector);
+    if (collector.expressionList.isEmpty()) {
+      return;
+    }
+    var issue = context.addIssue(functionDef.name(), MESSAGE);
+    collector.expressionList.forEach(call -> issue.secondary(call, SECONDARY_MESSAGE));
+  }
+
+  private static boolean isTfFunction(FunctionDef functionDefinition) {
+    return functionDefinition.decorators().stream()
+      .map(Decorator::expression).map(TreeUtils::getSymbolFromTree)
+      .filter(Optional::isPresent).map(Optional::get)
+      .map(Symbol::fullyQualifiedName)
+      .filter(Objects::nonNull)
+      .anyMatch("tensorflow.function"::equals);
+  }
+
+  private static class CallCollector extends BaseTreeVisitor {
+    private final Symbol originalSymbol;
+    List<Expression> expressionList = new ArrayList<>();
+
+    Set<FunctionDef> visited = new HashSet<>();
+
+    private CallCollector(@Nonnull Symbol originalSymbol) {
+      this.originalSymbol = originalSymbol;
+    }
+
+    @Override
+    public void visitCallExpression(CallExpression callExpression) {
+      Symbol symbol = TreeUtils.getSymbolFromTree(callExpression.callee()).orElse(null);
+      if (symbol == null) {
+        return;
+      }
+      if (symbol.equals(originalSymbol)) {
+        expressionList.add(callExpression.callee());
+      }
+      if (symbol.is(Symbol.Kind.FUNCTION)) {
+        symbol.usages().stream().filter(usage -> usage.kind() == (Usage.Kind.FUNC_DECLARATION)).forEach(usage -> usage.tree().parent().accept(this));
```

Comment:
I guess this (in combination with the `visited` list) works though I'm not 100% sure we care that much about indirect recursion (I think keeping it as an accepted FN would have been a good baby step) that we're going to visit all called functions for which we can find a definition. 

I'm not against it though as I guess the performance impact will be limited, given how narrow the scope of the rule is. Still, I'm a bit wondering whether this could cause FPs in some ways (with direct recursion, we can always assume a path is either dead code or can lead to the actual recursion - with indirect recursion, maybe the path that leads to the recursive code is neither dead code nor an actually possible path).